Middleware + Verify Session Cookies

johnny-t06 opened the pull request:
# Description
Prior to this, the middleware wasn't running because it was at the wrong file level. Firebase admin app did not follow a singleton pattern as recommended.
Also solved this with creation of verify session api + admin singleton.
![image](https://github.com/user-attachments/assets/70cf7e56-8a41-4f25-88fd-f3fbc2e71a7b)

Added verify session and log out session apis to handle cookies. 

There is a weird bug when signing in sometimes, you do not get redirected to `/private/course`

# Test
Log in and visit any routes, making sure you're not redirected to /login
Log out and make sure you can't visit any private routes.

johnny-t06 committed:
Verify Session api, logout api, middleware, env

nickbar01234 commented on firebaseAdmin.ts:
```diff
@@ -1,14 +1,40 @@
-if (!process.env.FIREBASE_ADMIN_SDK_JSON) {
-  throw new Error("FIREBASE_ADMIN_SDK_JSON is not set");
-}
+import "server-only";
```
Interesting library

nickbar01234 commented on package-lock.json:
```diff
@@ -45,143 +45,73 @@
-      "version": "7.24.7",
-      "resolved": "https://registry.npmjs.org/@babel/code-frame/-/code-frame-7.24.7.tgz",
-      "integrity": "sha512-BcYH1CVJBO9tvyIZ2jVeXgSIMvGZ2FDRvDdOIVQyuklNKSsx+eppDEBq/g47Ayw+RqNFE+URvOShmf+f/qwAlA==",
+      "version": "7.26.2",
```
How did you `package-lock` change but not your `package.json`? Am I missing something?

johnny-t06 replied:
Not sure, I might have uninstalled packages and deleted node_modules while testing

nickbar01234 replied:
Can you run `npm i` again locally? If you deleted packages, there should be a diff in `package.json`.

nickbar01234 commented on src/app/api/logout/route.client.ts:
```diff
@@ -0,0 +1,11 @@
+export const logOutCookie = async () => {
```
Why not simply `logout`? The implementation of using cookie should not be a factor.

nickbar01234 committed:
Persist session cookies (#42)

* Persist session cookie

* Remove log

* Push new page on authenticated

johnny-t06 committed:
Revert package-lock

johnny-t06 committed:
Rename logout api

johnny-t06 committed:
Rename log out api

johnny-t06 committed:
Package lock swc

johnny-t06 committed:
Fix error typo

nickbar01234 commented on firebaseAdmin.ts:
```diff
@@ -1,14 +1,40 @@
+      });
+    }
+  } catch (error) {
+    console.error("Error initializing Firebase Admin:", error);
```
Should just throw here since it's unrecoverable?

nickbar01234 commented on src/middleware.ts:
```diff
@@ -8,10 +7,20 @@ export const middleware = async (request: NextRequest) => {
   try {
-    // Verify the session cookie with Firebase Admin SDK
-    await admin.auth().verifySessionCookie(sessionCookie.value, true);
+    const res = await fetch(`${origin}/api/session`, {
```
Does `/api/session` by itself not work?

johnny-t06 replied:
[This](https://nextjs.org/docs/messages/middleware-relative-urls) mentions it a little. Since middleware runs on server, I think it doesn't have the relative url access.

nickbar01234 commented:
Given that we faced an infinite login earlier on, should we add any backdoor to get out of it in production? Some ideas

1. Mode to bypass requiring cookies to exist
2. If user is defined, but the middleware is rejected, should automatically logout to force the login flow again

nickbar01234 commented:
Otherwise, the changes look sensible

johnny-t06 committed:
Throw firebase admin, package lock, cookie set path

johnny-t06 committed:
Add bypass and get session api

johnny-t06 committed:
Frontend signout if middleware does not see session cookie

johnny-t06 commented on src/middleware.ts:
```diff
@@ -0,0 +1,48 @@
+    response.cookies.delete("session");
+  }
+  return response;
+};
```
On bad session token (i.e. DNE or not a valid token), we can reroute to /login?loggedOut=true. Then on /login and we check for loggedOut == true, to reset user context/trigger full logout.

johnny-t06 commented:
In my last two commits, I added a env to bypass the session check in our middleware. I also added a re-route to /login in the middleware that would fully reset the User Context. I added a GetSession api, but didn't use it. Let me know if I should remove it!

johnny-t06 committed:
Add error messages

johnny-t06 committed:
Same site, middleware no cache header
